lightning/src/ln/peer_handler.rs

```diff
@@ -1088,15 +1094,75 @@ impl<Descriptor: SocketDescriptor> PeerManager<Descriptor> {
 			}
 		};
 	}
+/// This function can be called every 30 seconds or so, it will:
+/// disconnect the Peer if awaiting_pong is true
+/// prepare to Ping all the Peers that the PeerManager is responsible for and set awaiting_pong to true
+///
+/// After calling this function one would want to call process_events() in order to finish Pinging the Peer
+/// Then one would want to call do_read_event() in order to reset awaiting_pong
+	pub fn timer_tick_occured(&self) {
+		let mut des_set: Vec<Descriptor> = Vec::new();
+		let mut peers_lock = self.peers.lock().unwrap();
+		let peers = peers_lock.borrow_parts();
+		
+		for (Descriptor, mut Peer) in peers.peers.iter_mut() {
```

Comment:
I think you can drop the des_set if you use HashMap.retain() instead - then you can move the disconnec logic into it and in place of the "peers.peers.remove()" call just return false to tell retain to drop the element.

Reply:
sweet, i'll check it out.
tbh the des_set stuff seemed like some sort of witchery.